refactor: monomorphizer public fields

Aristotelis2002 opened the pull request:
# Description

## Problem*

Resolves #9978 

This PR addresses the need to make Noir more extensible and library-friendly for external tools. Currently, it's not possible to interact with the monomorphization process from an external tool, which forces developers to maintain forks of Noir.

## Summary*

This PR makes the `Monomorphizer` struct and its relevant methods and fields public. This change will allow external tools, such as our formal verification tool, to integrate with the monomorphization process and use Noir as a library.

## Additional Context

This is one of the two PRs that will allow us to move our formal verification tool off of a Noir fork.

## Documentation*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

aakoshh commented on compiler/noirc_frontend/src/monomorphization/mod.rs:
```diff
@@ -72,16 +72,16 @@ pub(super) struct Monomorphizer<'interner> {
     /// but would also cause them to be re-evaluated which is a performance trap that would
     /// confuse users.
-    locals: HashMap<node_interner::DefinitionId, LocalId>,
+    pub locals: HashMap<node_interner::DefinitionId, LocalId>,
```
I'd consider adding read-only accessors to expose these without the ability to modify them, so that `Monomorphizer` can properly maintain its own data and invariants, while allowing 3rd parties to query it.

Aristotelis2002 commented:
Thanks again for pointing out the issues with the previous helper. We dropped it entirely and tightened the surface instead. Instead of adding a brand-new entry point, we made `queue_function` more granular and pulled parts of its body into a new public `queue_function_with_bindings`, letting external callers skip the `ExprId` path when they already have bindings and a location. We also made `monomorphize_debug` a lot more modular so downstream tooling can reuse the pieces it needs. Alongside that we exposed `follow_bindings` and added the few getters we need. You can see the exact result here: https://github.com/blocksense-network/noir/compare/refactor/monomorphizer-public-fields...refactor/monomorphizer-public-fields-v4. If you're happy with this direction we’ll update the PR branch to match and get it ready to merge.

jfecher reviewed (approved):
Apologies for the long review cycle - do you mind merging master to update the CI?
